casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala

```diff
@@ -299,57 +299,55 @@ class RuntimeManagerImpl[F[_]: Concurrent] private[rholang] (
       initHash: StateHash
   ): F[Either[(Option[DeployData], Failed), StateHash]] = {
 
-    def doReplayEval(
-        terms: Seq[InternalProcessedDeploy],
-        hash: Blake2b256Hash
-    ): F[Either[(Option[DeployData], Failed), StateHash]] =
-      Concurrent[F].defer {
-        terms match {
-          case processedDeploy +: rem =>
-            import processedDeploy._
-            for {
-              replayEvaluateResult <- replayComputeEffect(runtime)(hash, processedDeploy)
-              //TODO: compare replay deploy cost to given deploy cost
-              EvaluateResult(cost, errors) = replayEvaluateResult
-              cont <- DeployStatus.fromErrors(errors) match {
-                       case int: InternalErrors =>
-                         (deploy.some, int: Failed).asLeft[StateHash].pure[F]
-                       case replayStatus =>
-                         if (status.isFailed != replayStatus.isFailed)
-                           (deploy.some, ReplayStatusMismatch(replayStatus, status): Failed)
-                             .asLeft[StateHash]
+    def replaySingle(
+        hash: Blake2b256Hash,
+        processedDeploy: InternalProcessedDeploy
+    ): F[Either[(Option[DeployData], Failed), Blake2b256Hash]] = {
+      import processedDeploy._
+      for {
+        replayEvaluateResult <- replayComputeEffect(runtime)(hash, processedDeploy)
+        //TODO: compare replay deploy cost to given deploy cost
+        EvaluateResult(cost, errors) = replayEvaluateResult
+        cont <- DeployStatus.fromErrors(errors) match {
+                 case int: InternalErrors =>
+                   (deploy.some, int: Failed).asLeft[Blake2b256Hash].pure[F]
+                 case replayStatus =>
+                   if (status.isFailed != replayStatus.isFailed)
+                     (deploy.some, ReplayStatusMismatch(replayStatus, status): Failed)
+                       .asLeft[Blake2b256Hash]
+                       .pure[F]
+                   else if (errors.nonEmpty)
+                     hash.asRight[(Option[DeployData], Failed)].pure[F]
```

Comment:
This seems wrong.

Reply:
Why is that? What seems wrong?

Reply:
I guess I just don't understand the function well enough but it seems wrong that we return when errors are non-empty.

Reply:
Notice we're returning the previous initial hash. So the meaning of `processDeploy` is `try process, and return the resulting hash. If the execution was not successful, return the previous hash. We'll reset to it before continuing`. Notice also this is merely a refactoring of the direct recursion that took place before (IOW it maintains behavior).